flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/nn/QuadTree.scala

```diff
@@ -0,0 +1,352 @@
+ * The skeleton of the data structure was initially
+ * based off of the 2D Quadtree found here:
+ * http://www.cs.trinity.edu/~mlewis/CSCI1321-F11/Code/src/util/Quadtree.scala
+ *
+ * Many additional methods were added to the class both for
+ * efficient KNN queries and generalizing to n-dim.
+ *
+ * @param minVec vector of the corner of the bounding box with smallest coordinates
+ * @param maxVec vector of the corner of the bounding box with smallest coordinates
+ * @param distMetric metric, must be Euclidean or squareEuclidean
+ * @param maxPerBox threshold for number of points in each box before slitting a box
+ */
+class QuadTree(
+  minVec: Vector,
+  maxVec: Vector,
+  distMetric: DistanceMetric,
+  maxPerBox: Int) {
+
+  class Node(
+    center: Vector,
+    width: Vector,
+    var children: Seq[Node]) {
+
+    val nodeElements = new ListBuffer[Vector]
+
+    /** for testing purposes only; used in QuadTreeSuite.scala
+      *
+      * @return center and width of the box
+      */
+    def getCenterWidth(): (Vector, Vector) = {
+      (center, width)
+    }
+
+    /** Tests whether the queryPoint is in the node, or a child of that node
+      *
+      * @param queryPoint
+      * @return
+      */
+    def contains(queryPoint: Vector): Boolean = {
+      overlap(queryPoint, 0.0)
+    }
+
+    /** Tests if queryPoint is within a radius of the node
+      *
+      * @param queryPoint
+      * @param radius
+      * @return
+      */
+    def overlap(
+      queryPoint: Vector,
+      radius: Double): Boolean = {
+      val count = (0 until queryPoint.size).filter { i =>
+        (queryPoint(i) - radius < center(i) + width(i) / 2) &&
+          (queryPoint(i) + radius > center(i) - width(i) / 2)
+      }.size
+
+      count == queryPoint.size
+    }
+
+    /** Tests if queryPoint is near a node
+      *
+      * @param queryPoint
+      * @param radius
+      * @return
+      */
+    def isNear(
+      queryPoint: Vector,
+      radius: Double): Boolean = {
+      minDist(queryPoint) < radius
+    }
+
+    /**
+     * minDist is defined so that every point in the box
+     * has distance to queryPoint greater than minDist
+     * (minDist adopted from "Nearest Neighbors Queries" by N. Roussopoulos et al.)
+     *
+     * @param queryPoint
+     * @return
+     */
+    def minDist(queryPoint: Vector): Double = {
+      val minDist = (0 until queryPoint.size).map { i =>
+        if (queryPoint(i) < center(i) - width(i) / 2) {
+          math.pow(queryPoint(i) - center(i) + width(i) / 2, 2)
+        } else if (queryPoint(i) > center(i) + width(i) / 2) {
+          math.pow(queryPoint(i) - center(i) - width(i) / 2, 2)
+        } else {
+          0
+        }
+      }.sum
+
+      distMetric match {
+        case _: SquaredEuclideanDistanceMetric => minDist
+        case _: EuclideanDistanceMetric => math.sqrt(minDist)
+        case _ => throw new IllegalArgumentException(s" Error: metric must be" +
+          s" Euclidean or SquaredEuclidean!")
```

Comment:
`minDist` has the key property:  `minDist(queryPoint, node)` represents the distance between a point and a box, and has the property that for any point `P` in the box `Dist(queryPoint, P) < minDist(queryPoint, node)`. 

 I only know -- off the top of my head -- how to define such a function with that key property for the Euclidean/SquareEuclidean metric.  Not 100% sure how to generalize this for arbitrary metrics -- could very well be possible, but would have to think more about it.